L42-python-metadata-flags.md

```diff
@@ -0,0 +1,153 @@
+
+stub.initial_rpc_call(..., **important_call_options)
+stub.important_transaction_1(..., **important_call_options)
+stub.unimportant_transaction_2(..., **failfast_call_options)
+stub.important_transaction_3(..., **important_call_options)
+stub.unimportant_transaction_4(..., **failfast_call_options)
+```
+
+### Unsupported Design
+
+```Python
+# Channel level - Invalid Design
+channel = grpc.insecure_channel(..., wait_for_ready=True)
+stub = ...Stub(channel)
+
+stub.important_transaction_1(...)
+stub.important_transaction_2(...)
+stub.unimportant_transaction_3(...) # This will wait as well; user might get confused
+# --- A lot of code here ---
+stub.yet_another_transaction_500(...) # User might forget the channel setting
+```
+
+## Rationale
+
+### Default `None`, accept `bool` value
+* The `None` state defined that this value haven't been touched
+* It might be useful in some scenarios that interceptors wants to read its value and make changes
+
+### Whether to add channel level variable or not
+* The channel level variable is convenient and Pythonic
+* But it may create extra burden for future API evolution
+* **Please** comment about whether you feel like this design works for you or not
+* > Users "pay" for the complexity even if they don't use the feature.  --Eric Anderson
+* > We should be conservative when adding APIs, as we can easily add but cannot remove APIs. --Mehrdad Afshari
+
+### Why not integrate with `_channel_managed_call_management`?
+* Currently, we have a `managed_call` mechanism which works similar to call context in other languages.
+* It doesn’t work properly with flags.
+* It just pass a hard-coded `0` every time
+* It only be called for asynchronous calls. 
+* I prefer to implement it separately, with another private member variable.
+
+### Put the new variable in middle or tail
+* If this optional separate variable is right after mandatory variables, then developers can set it right away without typing extra variable name
+* But this will break if the existing program is using lazy position argument
+* So, put it in the tail sacrifices a little bit convenience to exchange the stability of API
+
+## Implementation
+
+Please refer to PR #16919
+
+### Testing Plan
+The testing will be similar to the unit test of _ChannelReadyFuture. It will contain three parts:
+
+1. Validate the options are set properly, and make sure the C-Core receives correct flags.
+2. Validate the behavior of client failed connect to an non-existing server without 'wait-for-ready' option.
+3. Validate the behavior of client successfully connect to a later-spawned server with 'wait-for-ready' option.
+
+### 'wait-for-ready' doesn’t cover all UNAVAILABLE status.
+As mentioned in the definition, during my experiments, the 'wait-for-ready' works to handle 'wait-for-ready' not 'wait-for-ready' which means if the server crashes while handling a request, the request will still be failed without retry or wait.
```

Comment:
I don't follow this statement: what does "works to handle 'wait-for-ready' not 'wait-for-ready'" mean?

Reply:
OMG, the editor's multi cursor has done some evil thing here...